README.md
[![Build Status](https://travis-ci.org/dmp1ce/decompose-drupal-7.svg?branch=master)](https://travis-ci.org/dmp1ce/decompose-drupal-7)

Drupal 7 environment for Decompose intended to help developers quickly start a Drupal 7 website. Because this project uses Docker instead of a virtual machine, the website is suitable for production as well as development and testing!

# Requirements

- [Decompose](https://github.com/dmp1ce/decompose)
- [Docker](http://www.docker.com/)
- [Docker Compose](http://docs.docker.com/compose/)

# Quick Start

``` sh
decompose --init https://github.com/dmp1ce/decompose-drupal-7.git
decompose build && decompose up
```
Then visit `http://localhost/install.php` to install Drupal. Username, password and host information for the database is already entered into the settings.php so all you'll need to do is setup admin user.

Modify the `containers/source/source/drupal/make.yml` file to add the modules, themes and libraries you want. After modifying `make.yml`, run `decompose build && decompose up` again.

# Usage

Decompose uses skeleton files in the `.decompose/environment/skel` to create a starting point for working on Drupal 7. Decompose will also initialize a git repository, so `git status` will show all of the files which have been created for you. Some of the files created are `.mo` files which are template files and contain elements (variables) like `{{PROJECT_NAME}}`. These elements are defined throughout the project but primarily in `elements` and `.decompose/elements`. Many tasks can be done with a decompose process such as `decompose build` or `decompose up`.

## File structure

After `decompose --init https://github.com/dmp1ce/decompose-drupal-7.git` is run, the following directory structure will be created.
```
├── containers
│   ├── backup
│   ├── backup_data
│   ├── data
│   ├── nginx
│   ├── nginx_proxy
│   ├── php
│   └── source
├── docker-compose.yml.mo
├── elements
└── processes
```
The containers directory is the configuration for all containers used by Drupal 7. The backup and backup_data containers are used for backing up Drupal 7 files and database settings using Duply. The data container stores Drupal uploaded files. The nginx container containers the nginx configuration for Drupal. The nginx_proxy directory is a special directory for configuration the nginx_proxy container which can host multiple websites from the same host! See jwilder/nginx-proxy. The php directory has the php-fpm and php configuration. Finally, the source directory is a special directory for sharing the Drupal source between nginx and php containers.

The `docker-compose.yml.mo` is a template file for creating the `docker-compose.yml` file. This file configures how each container interacts with each other. It also configures various settings such as environment variables, labels and restart settings.

The elements file contains non-private elements. Private elements such as passwords or API keys go in `.decompose/elements` and are not checked in. See [elements section](#elements) for details.

The processes file contains non-private processes. Private processes such as development only processes go in `.decompose/processes` and are not checked in. See [processes section](#processes) for details.

TODO: Go into more detail about each container which is used.

## Elements

Elements are essentually variables which can be used to change files generated from template files. Template files are any file with the extension `.mo`. Although, you can exclude `.mo` files from being processed using the `PROJECT_IGNORE_BUILD` element.

You can create your own elmeents by adding them to the `.decompose/elements` file if they are private and show not be checked in. If they are OK to be checked in and show to the public then you can add them to `elements` as long as `elements` is referenced by `.decompose/elments`.

### List of Elements

#### General Elements
- `PROJECT_ENVIRONMENT` : The current environment. Only `development` or `production` is supported.
- `PROJECT_NGINX_VIRTUAL_HOST` : The primary hostname used by nginx. This is what all other hosts will redirect to.
- `PROJECT_NGINX_VIRTUAL_HOST_ALTS` : All alternative hostnames. These are valid hostnames but they will be redirected to the `$PROJECT_NGINX_VIRTUAL_HOST` hostname. Multiple hostnames should be seperated by a space character. For example: `PROJECT_NGINX_VIRTUAL_HOST="example.com www.example.com"`
- `PROJECT_NGINX_DEFAULT_HOST` : Tells nginx_proxy where to go if it cannot find a matching domain. Be careful with this setting as it can override other projects which share the same nginx_proxy container.
- `PROJECT_SOURCE_PATH` : The relative location of the Drupal source from the source `Dockerfile`
- `PROJECT_SOURCE_HOST_PATH` : The relative location of the Drupal source from the `docker-compose.yml` file
- `PROJECT_VERSION_FILE` : The location and name of the version file to create on build
- `PROJECT_PHP_TIMEZONE` : The timezone of the server. Used to configures the PHP timezone.
- `PROJECT_DEBUG_NGINX` : Turn nginx debugging on
- `PROJECT_HTTP_SECURITY` : Turn on HTTP security. TODO: Is generating a htpasswd file necessary?
- `PROJECT_NGINX_PROXY_IMAGE` : Specify the nginx-proxy image to use. [jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy) is the default image.

#### SMTP Elements
- `PROJECT_SMTP_HOST` : SMTP host for sending email
- `PROJECT_SMTP_USER` : SMTP server user for seding email
- `PROJECT_SMTP_PASSWORD` : SMTP server password for sending email
- `PROJECT_SENDMAIL_PATH` : Setting PHP uses to send email

#### Database Elements
- `PROJECT_DB_USER` : Drupal database user
- `PROJECT_DB_DATABASE` : Drupal database name
- `PROJECT_DB_PASSWORD` : Drupal database user password
- `PROJECT_DB_ROOT_PASSWORD` : Drupal root user password

#### Backup Elements
- `PROJECT_BACKUP_GPG_KEY` : GPG key to use for encrypting backup
- `PROJECT_BACKUP_GPG_PW` : GPG password needed to use the key
- `PROJECT_BACKUP_TARGET` : Target location to send backup according to Duply documenation
- `PROJECT_BACKUP_CONFIG_TARGET` : Target location to send configuration backup. Unlike `PROJECT_BACKUP_TARGET` this uses scp scheme and not the Duply scheme.

#### Production Elements
- `PROJECT_PRODUCTION_SERVER_IP` : IP of production server
- `PROJECT_PRODUCTION_SERVER_USER` : User of production server which you connect to with SSH
- `PROJECT_PRODUCTION_SERVER_BASE_PATH` : Relative path from home which the project is located on production

#### Reverse Proxy Elements
- `PROJECT_WEBSITE_TO_EXPOSE_IP` : IP of machine which you want to expose from the reverse proxy server. `localhost` or the IP of a virtual machine most likely.
- `PROJECT_WEBSITE_TO_EXPOSE_PORT` : The port of the machine to expose from the reverse proxy server.
- `PROJECT_REVERSE_PROXY_USER` : The user used to SSH into the reverse proxy
- `PROJECT_REVERSE_PROXY_IP` : IP of the reverse proxy
- `PROJECT_REVERSE_PROXY_PORT` : The port of the reverse proxy which will map to the `PROJECT_WEBSITE_TO_EXPOSE_PORT`

TODO: grep for more, possibly hidden, elements

## Processes

Processes are typically common tasks that are needed when working on the project. Custom processes can be added to the `.decompose/processes` file. If you want to check in your processes then add them to `processes` and reference the file in `.decompose/processes`. See the `.decompose/processes` file for a simple example.

### List of Processes

`decompose --help` for list of processes

- `build` : Process template files and build Docker containers
- `up` : Start or restart Docker containers
- `import_db $db_file` : Import a Drupal database `.sql.gz` file.
- `explore_db` : Open a command prompt to MariaDB Drupal database
- `tail-dev-email` : Print to the screen the latest emails from a development environment
- `drush $command` : Run a drush command
- `backup-db` : Runs backup script. Also does backup on Drupal uploaded files, dispite the poor name
- `explore-php-container` : Opens a command prompt to the PHP container
- `env` : Shows if the current configuration is in production or development mode and shows URL host settings

### Additional Processes from decompose libraries
TODO: Instead of listing all of the processes provided by libraries, instead link to the library documentation when it is completed.

- `start_nginx_proxy` : Starts the nginx proxy container if it isn't already running
- `restart_nginx_proxy` : Restarts the nginx proxy container. If it isn't running then it will be started.
- `recreate_nginx_proxy` : Like restart but makes sure to create a new container instead of reusing an old one
- `start-reverse-proxy` : Starts a reverse proxy connection for exposing local environment to Internet. See element settings for `PROJECT_WEBSITE_TO_EXPOSE_IP`, `PROJECT_WEBSITE_TO_EXPOSE_PORT` and `│PROJECT_REVERSE_PROXY_*`. Also requires a properly configured SSH server to already be running.
- `update-production-server-to-latest` : Script for deploying currently checked in code to production server. See elements `PROJECT_PRODUCTION_SERVER_*`.
- `increment-tag` : Increments the last version tag by 1. For example v1.1 would be incremented to v1.2
- `ssh_production` : SSH into production server. Uses the same elmeents as `update-production-server-to-latest`.
- `generate_nginx_proxy_configs` : Generates the nginx proxy configuration files. See nginx_proxy container for details.
- `project-root` : Print project root directory
- `backup_config` : Backup all files which are not checked in. Important if for some reason the server is deleted and the configuration is needed to restore the site backup. See backup lib.
- `generate_gpg_backup_keys` : Generate GPG keys for backing up database and uploaded files.
- `generate_backup_server_ssh_access_key` : Generate SSH keys for accessing backup server to store backups
- `list-backups` : List all currently known backups. Useful for finding a backup to restore to.
- `restore-db` : Restore database from a backup
- `remove-untagged-docker-images` : Remove all untagged docker images. Useful for keeping Docker from using up all hard drive space.

# Local development

For developing modules or hacking on Drupal core, you'll probably want to be able to edit files and see the results right away on the website without needing to rebuild the project with `decompose build && decompose up`. To avoid needing to rebuild the project, create host volumes for the directory you are working in. The `source` container in the `docker-compose.yml.mo` file has a commented out section which is an example of how to create these volumes for Drupal core, custom modules, custom themes and/or libraries. If you also want to use the overrides in production, you'll probably want them copied to the source container at build time. See the `containers/source/Dockerfile.mo` for a commented section example.


CHANGELOG.md
# Changelog

Drupal 7 Decompose environment

## 2.1.1

- Add 'private' Drupal directory creation by default for 'source' container
- Minor change to `.decompose/elements` defaults for `NGINX_HOST`

## 2.1.0

- Have composer do two updates to get the modules updates when Drupal is updated
- Add process for updating `composer.lock` file

## 2.0.2

Initial version of changelog

- Try to make `decompose cron` return 0 when even though `drush` and `docker` want to return something else.


test/functions.bash
#!/bin/bash

# Requires DIR global variable which identifies script directory.

function setup_testing_environment() {
  echo "Setting up Docker testing environment ..."
  # Create dind daemon with a mount to project.
  echo "DinD service"
  testing_env_build=$(docker run --privileged --name decompose-docker-drupal7-testing -d docker:dind)
  [ ["$?" == "1" ] && echo "$testing_env_build" ]

  # Build testing image
  echo "Tester image"
  local project_directory=$(readlink -f "$DIR/../")
  local tmp_tester_build="/tmp/decompose-docker-drupal-7-testing"
  # Copy volume so we can safely dereference symlinks
  # Create docker container for doing tests
  cp -rL "$project_directory/." "$tmp_tester_build"
  cp "$project_directory/test/Dockerfile.tester" "$tmp_tester_build/Dockerfile"
  local testing_env_build=$(docker build -t decompose-docker-drupal-7-testing-tester "$tmp_tester_build/.")

  [ "$?" == "1" ] && echo "$testing_env_build"
}

function run_tests() {
  local tester_image="docker run --rm --link decompose-docker-drupal7-testing:docker decompose-docker-drupal-7-testing-tester"
  local return_code=0

  echo "Running BATS tests"
  echo "Drupal 7 tests"
  $tester_image bats "/app/skel/bats/drupal-7"
  local return_code+=$?

  #echo "Web library tests"
  #$tester_image bats "/app/skel/bats/web"
  #local return_code+=$?

  return $return_code
}

function teardown_testing_environment() {
  echo "Teardown Docker testing environment ..."
  echo "DinD service"
  testing_env_cleanup=$(docker rm -f decompose-docker-drupal7-testing)
  [ "$?" == "1" ] && echo "$testing_env_cleanup"

  echo "decompose build environment"
  testing_env_cleanup=$(docker rmi -f decompose-docker-drupal-7-testing-tester)
  [ "$?" == "1" ] && echo "$testing_env_cleanup"
}


test/bats/bats_functions.bash
#!/bin/bash

function echo_tmpdir() {
  # Save room in directory for uuid for saving test to view later.
  # Max size for a filename is 255 characters. UUID is 17 characters plus '-' character.
  echo "$BATS_TMPDIR/${BATS_TEST_NAME:0:216}"
}

function setup_testing_environment() {
  # Setup project in temporary directory.
  local tmpdir=$(echo_tmpdir)
  
  # Directory for setting up tests
  local tmpdir=$(echo_tmpdir)
  mkdir -p "$tmpdir/build-test-environment"
 
  # Setup build test environment
  cp -r "$BATS_TEST_DIRNAME/../../." "$tmpdir/build-test-environment"

  # Remove git submodules and reinitialize git because moving .git repositories
  # is problematic:
  # https://stackoverflow.com/questions/17568543/git-add-doesnt-work/17747571#17747571 
  mv "$tmpdir/build-test-environment/.git" "$BATS_TMPDIR/$(uuidgen)"
  mv "$tmpdir/build-test-environment/.gitmodules" "$BATS_TMPDIR/$(uuidgen)"
  find "$tmpdir/build-test-environment/" -type f -name .git -exec rm {} \;
  
  local git_url=$(realpath $tmpdir/build-test-environment)
  git -C "$git_url" init
  git -C "$git_url" config user.email "tester@example.com"
  git -C "$git_url" config user.name "tester"
  git -C "$git_url" add .
  git -C "$git_url" commit -m "Initial commit"

  # Make and set home directory
  mkdir -p "$tmpdir/user-home"
  export HOME=$(realpath "$(echo_tmpdir)/user-home")

  # Set current working directory
  export WORKING="$tmpdir/build-test"
  mkdir -p "$tmpdir/build-test"

  # Init build and initialize environment
  local error_output=$(export HOME=$HOME && cd $WORKING && \
    decompose --init "$git_url" 2>&1 1>/dev/null)
  if [ -n "$error_output" ]; then
    echo "'decompose --init' had errors"
    echo "$error_output"
    return 1
  fi

  # Reference elements and processes files in WORKING directory
  cp "$BATS_TEST_DIRNAME/fixtures/elements" "$tmpdir/build-test/.decompose"
  #cp "$BATS_TEST_DIRNAME/fixtures/processes" "$tmpdir/build-test/.decompose"
}

function teardown_testing_environment() {
  # TODO: Cean up generated volumes

  # Directory for setting up tests
  local tmpdir=$(echo_tmpdir)

  # Move generated test files to random temporary
  # directory where it will eventually be cleaned up
  mv "$tmpdir" "$tmpdir-$(uuidgen)"
}

# vim:syntax=sh tabstop=2 shiftwidth=2 expandtab


test/decompose_environment/docker-entrypoint.sh
#!/bin/sh
set -e

#if [ "${1:0:1}" = '-' ]; then
#	set -- docker "$@"
#fi

# if our command is a valid Docker subcommand, let's invoke it through Docker instead
# (this allows for "docker run docker ps", etc)
#if docker "$1" --help > /dev/null 2>&1; then
#	set -- docker "$@"
#fi

# if we have "--link some-docker:docker" and not DOCKER_HOST, let's set DOCKER_HOST automatically
if [ -z "$DOCKER_HOST" -a "$DOCKER_PORT_2375_TCP" ]; then
	export DOCKER_HOST='tcp://docker:2375'
fi

exec "$@"


skel/bats/drupal-7/bats_functions.bash
#!/bin/bash

function echo_tmpdir() {
  echo "$BATS_TMPDIR/$BATS_TEST_NAME"
}

function setup_testing_environment() {
  # Setup project in temporary directory.
  local tmpdir=$(echo_tmpdir)
  
  # Directory for setting up tests
  local tmpdir=$(echo_tmpdir)
  mkdir -p "$tmpdir/build-test-environment"
 
  # Setup build test environment
  cp -r "$BATS_TEST_DIRNAME/../../../." "$tmpdir/build-test-environment"

  # Remove git submodules and reinitialize git because moving .git repositories
  # is problematic:
  # https://stackoverflow.com/questions/17568543/git-add-doesnt-work/17747571#17747571 
  mv "$tmpdir/build-test-environment/.git" "$BATS_TMPDIR/$(uuidgen)"
  mv "$tmpdir/build-test-environment/.gitmodules" "$BATS_TMPDIR/$(uuidgen)"
  find "$tmpdir/build-test-environment/" -type f -name .git -exec rm {} \;
  
  local git_url=$(realpath $tmpdir/build-test-environment)
  git -C "$git_url" init
  git -C "$git_url" config user.email "tester@example.com"
  git -C "$git_url" config user.name "tester"
  git -C "$git_url" add .
  git -C "$git_url" commit -m "Initial commit"

  # Make and set home directory
  mkdir -p "$tmpdir/user-home"
  export HOME=$(realpath "$(echo_tmpdir)/user-home")

  # Set current working directory
  export WORKING="$tmpdir/build-test"
  mkdir -p "$tmpdir/build-test"

  # Init build and initialize environment
  local error_output=$(export HOME=$HOME && cd $WORKING && \
    decompose --init "$git_url" 2>&1 1>/dev/null)
  if [ -n "$error_output" ]; then
    echo "'decompose --init' had errors"
    echo "$error_output"
    return 1
  fi

  # Reference elements and processes files in WORKING directory
  cp "$BATS_TEST_DIRNAME/fixtures/elements" "$tmpdir/build-test/.decompose"
  #cp "$BATS_TEST_DIRNAME/fixtures/processes" "$tmpdir/build-test/.decompose"
}

function teardown_testing_environment() {
  # Directory for setting up tests
  local tmpdir=$(echo_tmpdir)

  # Move generated test files to random temporary
  # directory where it will eventually be cleaned up
  mv "$tmpdir" "$tmpdir-$(uuidgen)"
}

# vim:syntax=sh tabstop=2 shiftwidth=2 expandtab
